### ISSUE TITLE
Make caching optional?

### ISSUE THREAD

**Issue by hynek** [hynek/svcs#12]
Should there be a way to force a new value either at factory level or at `get()` level – or both?

**Comment by elfjes** [comment]
I would say this is primarily a factory-concern. Use case that springs to mind is have a (connection) pool where you have a pool in your container and can `with container.get(Connection):` where the factory for `Connection` obtains one from the pool. as a consumer, you don't want be bothered with details of whether it should be cached or not.

So perhaps both? "As a consumer of a service, I want to be in control of the life time of that service"... nothing springs to mind, but perhaps that is my (lack of) imagination

**Comment by hynek** [comment]
to clarify, the caching i'm speaking about is that if I ask for a connection, it is stored in the container and returned to everyone who asks for it too. the factory is _not_ involved at all.

this is usually what you want – at least I haven't run into examples that want the opposite. it can be problematic to have two database connections for example and try to combine them into one transaction.

---

to make this the job of the factory, the factory would need the container. Even if we keep the short path within container, you'd have to write something like:

```python
def factory(svcs_container):
    x = create()
    svcs.container.store(x)
    return x
```

Which would make it the wrong thing by default, which is not what I want.

I _could_ live with having store/cache on the container, but only require it when the user called `register_factory(Foo, foo, cache=False)` and opts into taking care of that themselves.

But as said, I have not really needed this yet, so I think it's better to wait for people yell at me for something like this.

**Comment by elfjes** [comment]
Right, I didn't express myself correctly. What I meant is that in general it is at registration time that you'd want to specify whether containers should cache an instance:

```python
registry.register_factory(int, lambda: 42, cache=False)
```

and not at request time:

```python
container.get(int, cache=False)
```

I can think of use cases for the first (see previous reply), but not for the latter

**Comment by hynek** [comment]
Ah right!

TRBF my uses for the latter would be mostly hacky things in testing. But let's wait and see for now.
